Detectors/MUON/MID/Raw/src/ROBoardConfigHandler.cxx

```diff
@@ -0,0 +1,181 @@
+  if (cfgIt == mROBoardConfigs.end()) {
+    throw std::runtime_error(fmt::format("Cannot find local board {:02x}", uniqueLocId));
+  }
+  return cfgIt->second;
+}
+
+bool ROBoardConfigHandler::load(const char* filename)
+{
+  std::ifstream inFile(filename);
+  if (!inFile.is_open()) {
+    return false;
+  }
+  std::vector<ROBoardConfig> configurations;
+  std::string line, token;
+  while (std::getline(inFile, line)) {
+    if (std::count(line.begin(), line.end(), ' ') < 5) {
+      continue;
+    }
+    if (line.find('#') < line.find(' ')) {
+      continue;
+    }
+    ROBoardConfig cfg;
+    std::stringstream ss;
+    ss << line;
+    std::getline(ss, token, ' ');
+    cfg.boardId = static_cast<uint8_t>(std::strtol(token.c_str(), nullptr, 16));
+    std::getline(ss, token, ' ');
+    cfg.statusWord = static_cast<uint32_t>(std::strtol(token.c_str(), nullptr, 16));
+    for (int ich = 0; ich < 4; ++ich) {
+      std::getline(ss, token, ' ');
+      auto mask = static_cast<uint32_t>(std::strtol(token.c_str(), nullptr, 16));
+      cfg.masksBP[ich] = (mask >> 16) & 0xFFFF;
+      cfg.masksNBP[ich] = (mask & 0xFFFF);
+    }
+    configurations.emplace_back(cfg);
+    inFile.close();
+  }
+  set(configurations);
+  return true;
+}
+
+void ROBoardConfigHandler::write(const char* filename) const
+{
+  /// Writes the masks to a configuration file
+  std::vector<ROBoardConfig> configs;
+  for (auto& cfgIt : mROBoardConfigs) {
+    configs.emplace_back(cfgIt.second);
+  }
+
+  std::sort(configs.begin(), configs.end(), [](const ROBoardConfig& cfg1, const ROBoardConfig& cfg2) { return cfg1.boardId < cfg2.boardId; });
+
+  std::ofstream outFile(filename);
+  for (auto& cfg : configs) {
+    outFile << cfg << std::endl;
+  }
+  outFile.close();
+}
+
+void ROBoardConfigHandler::set(const std::vector<ROBoardConfig>& configurations)
+{
+  mROBoardConfigs.clear();
+  for (auto& cfg : configurations) {
+    mROBoardConfigs.emplace(cfg.boardId, cfg);
+  }
+}
+
+void ROBoardConfigHandler::updateMasks(const std::vector<ROBoard>& masks)
+{
+  for (auto& mask : masks) {
+    auto cfgIt = mROBoardConfigs.find(mask.boardId);
+
+    // First we check if some patterns has zeros.
+    // When set xORy for boards with no Y input.
+    // So in this case we explicitly mask Y.
+    bool isMasked = ((cfgIt->second.statusWord & crateconfig::sXorY) != 0);
+    for (int ich = 0; ich < 4; ++ich) {
+      if (mask.patternsBP[ich] != 0xFFFF || mask.patternsNBP[ich] != 0xFFFF) {
+        isMasked = true;
+      }
+    }
+
+    if (isMasked) {
+      cfgIt->second.statusWord |= crateconfig::sMonmoff;
+      cfgIt->second.masksBP = mask.patternsBP;
+      if ((cfgIt->second.statusWord & crateconfig::sXorY) == 0) {
+        cfgIt->second.masksNBP = mask.patternsNBP;
+      }
+    }
+  }
+}
+
+std::vector<ROBoardConfig> makeDefaultROBoardConfig(uint16_t gbtUniqueId)
+{
+  // FIXME: in the current configuration, when an Y strip covers several local boards
```

Comment:
is there a JIRA ticket corresponding to this (rather important) FIXME ?

Reply:
Not sure to which JIRA project this belongs. It is just an internal decision. Also, since the Y signal is physically not sent to some boards, we will have to leave with almost half of the detector in X OR Y. So, for consistency and to avoid biases we might even decide to keep X OR Y by default for all boards.

Reply:
> Not sure to which JIRA project this belongs

Good point. As it's an MID-wide thing, one option would be to use the run coordination project (R3C ; MID component) ?